### ISSUE TITLE
Failed to create schema with optional fields

### ISSUE THREAD

**Issue by Gr1N** [samuelcolvin/pydantic#213]
Just tried to use schema creation and it's awesome, but I found, I think, bug with optional fields.

Here is my model:
```
from typing import Optional
from pydantic import BaseModel

class TestModel(BaseModel):
    foo: Optional[int] = None

print(TestModel.schema_json(indent=2))
```

Output:
```
Traceback (most recent call last):
  File "test_pydantic_schema.py", line 9, in <module>
    print(TestModel.schema_json(indent=2))
  File "/pydantic/pydantic/main.py", line 288, in schema_json
    return json.dumps(cls.schema(by_alias=by_alias), default=pydantic_encoder, **dumps_kwargs)
  File "/pydantic/pydantic/main.py", line 278, in schema
    s['properties'] = {f.alias: f.schema(by_alias) for f in cls.__fields__.values()}
  File "/pydantic/pydantic/main.py", line 278, in <dictcomp>
    s['properties'] = {f.alias: f.schema(by_alias) for f in cls.__fields__.values()}
  File "/pydantic/pydantic/fields.py", line 146, in schema
    if issubclass(self.type_, Enum):
TypeError: issubclass() arg 1 must be a class
```

Same issue with `Union`:
```
from typing import Union
from pydantic import BaseModel

class TestModel(BaseModel):
    foo: Union[str, int]

print(TestModel.schema_json(indent=2))
```

I think `pydantic` should cover cases like this, I don't know what is the best way to solve issue with `Optional`, but in case of `Union` `oneOf` or `anyOf` rule can be used http://json-schema.org/latest/json-schema-validation.html#rfc.section.6.7.

**Comment by samuelcolvin** [comment]
yes.

I guess `Optional` is just equivalent of `Union[None, Foobar]`.

As discussed on #129 I'm not that bothered by `json-schema` or it's equivalents, but yes, something like that makes sense.

Perhaps we just have `"type": ["int", "float"]` and an array for `type` just means "union"?

**Comment by Gr1N** [comment]
For JSON Schemas, "union" means:
```
{
    "type": "object",
    "properties": {
        "foo": {
            "anyOf": [
                { "type": "int },
                { "type": "float" },
            ]
    }
}
```

**Comment by Gr1N** [comment]
Also I don't think that we need to use `anyOf` for `Optional`, I think better to convert this case to property with `default=null`.

**Comment by x1ah** [comment]
May missed `enum` type ?

In my opinion, we should carefully change the field structure. and `typing.Union` 's field info looks good.

```python
from typing import Union
from pydantic import BaseModel


class Foo(BaseModel):
    ubar = Union[int, str]

print(Foo.__fields__['r'].info)
# {'type': 'typing.Union[int, str]',
#  'required': True,
#  'sub_fields': [<Field r_int: type='int', required=True, validators=['int_validator']>,
#   <Field r_str: type='str', required=True, validators=['not_none_validator', 'str_validator','anystr_strip_whitespace', 'anystr_length_validator']>]}
```

and similar for `List[int, str]`, we get

```python
{'type': 'typing.List[int]',
 'required': True,
 'sub_fields': [<Field r_int: type='int', ...>]}
```

by this way, this problem will be solved without changing the original fields struct, just put into `sub_fields`.

What do you think?

**Comment by samuelcolvin** [comment]
Just copying python's syntax is ok with me, it'll make it slightly harder for people to read programmatically, but it'll be more concise and extendible.

The case of enums is already implemented via `choices`.

**Comment by Gr1N** [comment]
> are their any case's I've missed?

What about `Optional[any type]`?

And one question, why we want to invent new format? Why not use JSON Schema spec? I understand that spec in some cases can be very explicit, but it covers all needs to my mind. Also it's good to follow specification in cases if we want to generate API specs from `pydantic` models and give them to another team who works, for example, with different language, and if generated specs follow JSON schema spec we can use them to generate code for any language we want.

**Comment by samuelcolvin** [comment]
I think your suggestion of `default=null` for `Optional` is good.

> And one question, why we want to invent new format? Why not use JSON Schema spec?

see https://github.com/samuelcolvin/pydantic/issues/129#issuecomment-364754565, also:
1. It would be masses of work (perhaps more than the rest of pydantic combined) to fully support JSON schema, which is only draft anyway.
1. pydantic isn't tied to JSON and I think we would run into lots of conflicts, eg. (AFAIK) json schema doesn't cover lots of types with pydantic supports, eg. decimal, uuid, dict keys have to be strings in json

**Comment by Gr1N** [comment]
Understand. Can you please share how cases `Union[int, FoobarModel]` and `Union[FooModel, BarModel]` will be looks in proposed format?

**Comment by Gr1N** [comment]
LGTM, let's try to go with your proposal.

**Comment by samuelcolvin** [comment]
as per discussion on gitter:
* the top level property should change `type` > `schema`
* there should be a shortened display of just `'schema': {'type': 'int'}` which is just `'schema': 'int'`, this applies both at the top level for simple types and where simple types are nested within more complex types.

Overal the design is encapsulated by:

```py
'foobar': {
  'schema': {
    'mode': 'list',
    'item_type': {
      'mode': 'any_of',
      'types': [
        'float', 'int',
      ],
    }
  },
  'name': 'Foobar',
  'required': True,
}
```

**Comment by samuelcolvin** [comment]
I've implemented this in #232, it's not exactly as described above, but similar. The nice thing is that this doesn't break much of what's already implemented in `schema()`
